features/0337-jwe-compliant-envelope/README.md

```diff
@@ -0,0 +1,239 @@
+```json
+    {
+    "protected": base64url({
+        "typ": "prs.hyperledger.aries-auth-message",
+        "alg": "ECDH-SS+A256KW",
+        "enc":"XC20P"
+    }),
+    "recipients": [
+        {
+            "encrypted_key": "base64url(encrypted CEK)",
+            "header": {
+                "apu": "base64url(random_bytes(64))",
+                "nonce": "base64url(random_bytes(nonce_size))",
+                "tag": "base64url(encrypted CEK tag)",
+                "kid": "base58_encode(Recipient enckey)",
+                "spk": "EphemeralJWK(Sender enckey, Recipient enckey)",
+            }
+        },
+        {
+            "encrypted_key": "base64url(encrypted CEK)",
+            "header": {
+                "apu": "base64url(random_bytes(64))",
+                "nonce": "base64url(random_bytes(nonce_size))",
+                "tag": "base64url(encrypted CEK tag)",
+                "kid": "base58_encode(Recipient enckey)",
+                "spk": "EphemeralJWK(Sender enckey, Recipient enckey)",
+            }
+        }
+    ],
+    "aad": "base64url(sha256(concat('.',sort([base58(recipients[0] enckey), ..., base58(recipients[n] enckey ])))))",
+    "iv": "base64url(content encryption IV)",
+    "ciphertext": "base64url(C20P(DidExchange payload, base64Url(json($protected)+'.'+$aad), content encryption IV, CEK))"
+    "tag": "base64url(AEAD Authentication Tag)"
+}
+```
+With `spk` being an embedded compact JWE defined as the following 5 components:
+```bash
+base64Url(Headers).
+base64Url(encryptedKEK).
+base64Url(random_bytes(nonce_size)).
+base64Url(encryptedJWK cipher).
+base64Url(encryptedJWK tag)
+```
+where `encryptedJWK` raw format is the sender's public encryption (static) key as JWK:
+```json
+ [
+     "kty": "OKP",
+	 "crv": "X25519",
+	 "x": base64url(Sender enckey),
+ ]
+```
+and `Headers` being the following JSON:
+```json
+ [
+	 "typ":  "jose",
+	 "cty": "jwk+json",
+	 "alg": "ECDH-ES+A256KW",
+	 "enc": "XC20P",
+	 "epk": [
+	      "kty": "OKP",
+	      "crv": "X25519",
+	      "x": base64url($epk), // this is an ephemeral public key created to encrypt KEK
+	 ],
+	 "iv": base64url($knonce), // IV used to encrypt KEK
+	 "tag": base64url($ktag), // tag of encyprtedKEK
+ ]
+```
+So the encrypted sender encryption (static) key is encrypted into a JWE envelope stored in `spk`. Notice this envelope is encrypted using an `epk` where key wrapping is in ES mode as stated in `alg` header field.
+
+
+### Fully FIPS compliant content encryption and SS key wrapping mode (proposed)
+The fully FIPS compliant example is the same as the XC20P example above with the exception of the `enc` header field in both the outer JWE envelope and the inner sender envelope will be `A256GCM` instead of `XC20P`
+
+
+These are proposed formats. The RFC will be updated once community consensus is made.
+
+## Prior art
+
+- The [JWE](https://tools.ietf.org/html/rfc7518) family of encryption methods.
```

Comment:
JOSE ECDH-1PU draft could be helpful: https://tools.ietf.org/html/draft-madden-jose-ecdh-1pu-02

Reply:
this is definitely helpful for Authcrypt, but it's only a draft..

Reply:
There is a companion blog post: https://neilmadden.blog/2018/11/26/public-key-authenticated-encryption-and-why-you-want-it-part-ii/

> a NIST-approved key-agreement mode that achieves PKAE

(PKAE = public key authenticated encryption)
See https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-56Ar3.pdf

Reply:
updated RFC to include ECDH-1PU encryption for Authcrypt according to this IETF draft.